Add benchmark CI

grief8 opened the pull request:
This PR adds a benchmark workflow to the Asterinas repository, designed to assess performance metrics for the project. This automated workflow is scheduled to run at midnight every day.

The workflow executes on the latest Ubuntu runner and is constrained to a 60-minute time limit to optimize resource consumption. It utilizes a predefined Docker container, `asterinas/asterinas:0.4.2`, to provide a consistent testing environment across workflow runs.

The benchmark process begins with a standard checkout step, ensuring the latest code changes are retrieved. It then installs `jq`, a lightweight and flexible command-line JSON processor, which is essential for manipulating the JSON outputs generated by the benchmark scripts. I recommend adding it to the container image in the later environment.

Two key benchmarks are performed:

1. The `sysbench` benchmark, which stress-tests the system and provides metrics on its performance. The results are stored in a JSON file called `result_sysbench.json`.

2. The `getpid` benchmark, which measures the time taken to execute the `getpid` system call. The results from this benchmark are stored in a second JSON file named `result_getpid.json`.

Both results are processed by the `benchmark-action/github-action-benchmark@v1`, a GitHub Action tailored for handling benchmark data. This action supports a custom tool option, `customSmallerIsBetter`, implying that lower values are indicative of better performance for both benchmarks.

Benchmark results are stored in the `stable/benchmarks` directory and are automatically pushed to the `gh-pages` branch, effectively updating the project's GitHub Pages with the latest performance data. 
**Note that an empty branch `gh-pages` should be created to store the data before applying this PR.**

The workflow is configured to raise an alert if the benchmarks detect a potential performance regression that exceeds a 25% threshold for `getpid` and 30% threshold for `sysbench`. Commenting on alerts is enabled, providing immediate feedback on the commit responsible for performance degradation. Additionally, if an alert is triggered, the entire workflow fails, which acts as a blocking signal to developers, emphasizing that a serious performance issue needs to be addressed.

tatetian commented:
May I see the visualized results of the benchmark?

> a potential performance regression that exceeds a 200% threshold.

What is the meaning of 200% threshold? The latency of getpid increases from 100ns to 200ns? If so, this threshold seems to be too generous.

tatetian commented:
@cchanging Could you review this PR and provide your valuable feedbacks? Thanks.

cchanging commented on Makefile:
```diff
@@ -36,6 +37,14 @@ else ifeq ($(AUTO_TEST), boot)
+ifeq ($(BENCHMARK), sysbench)
+CARGO_OSDK_ARGS += --init-args="/benchmark/sysbench.sh"
+else ifeq ($(BENCHMARK), iperf3)
+CARGO_OSDK_ARGS += --init-args="/benchmark/iperf3.sh"
```
Add this parameter after the iperf benchmark CI has actually been implemented. Currently it's useless and we don't even have `iperf3.sh`.

grief8 replied:
Thank you for reminding! I will remove it. Actually we have iperf3 toolkit inside Asterinas, but it isn't supported as #866. We can add it after the problem is solved.

cchanging commented on regression/benchmark/result_template_sysbench.json:
```diff
@@ -0,0 +1,26 @@
+        "unit": "ms",
+        "value": 0,
+        "extra": "max"
+    },
```
So, are we going to compare all the recorded results here? I think outcomes like `max`, `min` might be somewhat random, even if you've set a relatively large threshold. I suggest that comparing the `avg` results would be sufficient for now.

grief8 replied:
For current version, the answer is yes. `sysbench` gives `max`, `min`, `avg` and `per95`. Maybe the latter two are more suitable to be set up for comparison.

cchanging commented on regression/benchmark/benchmarks.sh:
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
```
Can add a blank line here.

cchanging reviewed (commented):
Thank you for adding the benchmark CI capability, which is very helpful for us to observe the impact of new PRs on performance. Currently, this PR has set up the framework and supports testing `sysbench` and `getpid` in CI. Overall, I believe that considering we may add more benchmarks in CI in the future, we could store all the test results in a single result file, or use some other form of encapsulation. This way, we won't need to add too many fields to `benchmarks.yml` for each new benchmark (currently each benchmark operation is a separate section, which can be redundant).

I also suggest incorporating `membench` or a similar basic memory test, as this is an important performance metric as well. Regarding the threshold, I'm quite curious about the level of performance fluctuation measured on GitHub CI machines, which I believe will significantly influence our setting of the threshold.

grief8 commented:
> May I see the visualized results of the benchmark?
> 
> > a potential performance regression that exceeds a 200% threshold.
> 
> What is the meaning of 200% threshold? The latency of getpid increases from 100ns to 200ns? If so, this threshold seems to be too generous.

There is an example in https://grief8.github.io/asterinas/stable/benchmarks/ without abundant data. The 200% threshold may be too generous to be triggered. I will adjust it to a more reasonable one, e.g., 50%, after I get enough CI results on my repository.

grief8 commented:
Another issue is that Github usually uses the `gh-pages` branch to save data and display web pages, so I followed this setting, but if the `benchmarks` branch is more readable, I can also modify the workflow file.

tatetian commented:
> I will adjust it to a more reasonable one, e.g., 50%, after I get enough CI results on my repository.

Ideally, we should be warned about any performance regression. For every captured case of performance regression, we should analyze the source of the performance degradation and decide if it is due to the necessary cost introduced by a new feature or an suboptimal implementation that should be fixed.

Assuming the precision or standard error or standard deviation of a measured result is `D%`, if the overhead increases beyond `(2 * D)%`, then we are most likely witnessing a performance degradation.

grief8 commented:
> > I will adjust it to a more reasonable one, e.g., 50%, after I get enough CI results on my repository.
> 
> Ideally, we should be warned about any performance regression. For every captured case of performance regression, we should analyze the source of the performance degradation and decide if it is due to the necessary cost introduced by a new feature or an suboptimal implementation that should be fixed.
> 
> Assuming the precision or standard error or standard deviation of a measured result is `D%`, if the overhead increases beyond `(2 * D)%`, then we are most likely witnessing a performance degradation.

This is a very reasonable algorithm. 

Considering that we provide a separate branch in CI for PRs to just compare performance, we may be able to set a more strict threshold for PRs, such as `1.5*D%`, to ensure that CI is less likely to fail after it is merged.

grief8 commented:
Benchmark on pull requests is not a good idea because it may not rely on the latest commits.

grief8 commented:
The latest version introduces the comparison between Linux and Asterinas by replacing the old benchmark action with my own action. The renderings can be seen on https://grief8.github.io/asterinas/stable/benchmarks/.

However, I think there still remains some issues to be decided.
- If we want to eliminate the effect of QEMU, we'd better set up a Linux VM for our CI. If we choose a pre-built image, where should it place? If we construct such an image every time, is the overhead too high?
- Should we keep the benchmark CI running on GitHub? (KVM is not supported on GitHub CI.)

grief8 commented:
Add lightweight Linux-5.15.0 kernel and reuse initramfs of Asterinas as our Linux baseline VM.

grief8 commented:
Although I added the proxy on my CI machine, the download of `vdso` was still not particularly stable, so I introduced `nick-invision/retry@v2` to retry the failed compilation of asterinas. Now the success rate is higher. 

However, I still strongly recommend persisting vdso and vmlinuz files to container images. And additional path check should be made in the Makefile to ensure that someone is able to run asterinas outside the container.

grief8 commented:
Bump container image version to 0.5.0.

tatetian commented on regression/benchmark/init:
```diff
@@ -0,0 +1,22 @@
+#!/bin/sh
```
The file name "init" is not quite a descriptive one

tatetian commented on regression/benchmark/benchmarks.sh:
```diff
@@ -0,0 +1,103 @@
+    if [ ! -d "${kernel_dir}" ]; then
+        mkdir -p "${kernel_dir}"
+    fi
+    local linux_kernel="${kernel_dir}/vmlinuz-5.15.0-105-generic"
```
Update the Dockerfile to include this kernel image.

```
        curl -L -o "${linux_kernel}" \
            -H "Accept: application/vnd.github.v3.raw" \
            "https://api.github.com/repos/grief8/storage/contents/vmlinuz-5.15.0-105-generic"
```

Use an official source to download the image. Not your own repository.

tatetian commented on .github/workflows/benchmarks.yml:
```diff
@@ -0,0 +1,53 @@
+        on_retry_command: make clean
+
+    - name: Store benchmark results
+      uses: grief8/github-action-benchmark@v1
```
Is it possible to move this action to Asterinas org? And this action---its name and arguments--does not seem to be very self-describing. For example, the name is "github-action-benchmark", but apparently, the job is not responsible for running the benchmark, only collecting and visualizing the results. So it should use a more accurate name.

tatetian commented on .github/workflows/benchmarks.yml:
```diff
@@ -0,0 +1,53 @@
+    runs-on: ${{ matrix.runner }}
+    strategy:
+      matrix:
+        runner: [ubuntu-latest, self-hosted]
```
Does this mean that the benchmarks are going to be run on both Github and self-host CI? But we do not need to run on Github CI, whose results are not accurate.

tatetian commented on .github/workflows/benchmarks.yml:
```diff
@@ -0,0 +1,53 @@
+        chmod +x regression/benchmark/benchmarks.sh
+        git config --global --add safe.directory /__w/asterinas/asterinas
+        git config --global http.sslVerify false
+        git config --global http.version HTTP/1.1
```
What is the purpose of these unconventional setup? Better explain the surprising code with comments.

tatetian reviewed (commented):
Good job on enabling benchmark CI! The result diagrams are visually appealing. I love how our benchmark compares Asterinas with Linux head-to-head.

By adding the benchmark directory, it no longer makes sense to call the parent directory "regression". I think it should be renamed to "test". We did not choose to use the name "test" because Cargo treats a "test" directory specially. Now we have OSDK. This should not be an issue. So I think it is now a good time to rename "regression" to "test".

grief8 commented on .github/workflows/benchmarks.yml:
```diff
@@ -0,0 +1,53 @@
+name: Benchmarks Test
+on:
+  schedule:
+    - cron: '0 0 * * *'
```
Certainly! The line is a cron expression means "schedule it to 0:00 every day". I will add an explanation later. 
The detailed rules are as follows:
```log
* * * * *
- - - - -
| | | | |
| | | | ----- Day of the week (0 - 7) (Sunday=0 or 7)
| | | ------- Month (1 - 12)
| | --------- Day of the month (1 - 31)
| ----------- Hour (0 - 23)
------------- Minute (0 - 59)
```

grief8 commented:
Current commits make the following changes:

- Optimize benchmark structure, and put each benchmark in a separate folder
- Change threshold to be fetched from the configuration file
- Modify Dockerfile to persist vdso and Linux kernel files

Note: OSDK CIs seems to encounter some problems, which are not caused by this PR.

tatetian commented on regression/benchmark/benchmarks.sh:
```diff
@@ -0,0 +1,125 @@
+
+    # Copy sysbench
+    echo "Copying sysbench to /benchmark..."
+    cp /usr/local/benchmark/sysbench/bin/sysbench /benchmark
```
The current code organization still fails to keep benchmark-specific code such as the one specific to `sysbench` under its own directory. This script still hardcodes information about `sysbench` and `getpid`.

Similarly, `benchmark/config.json` is also a centralized place that keep information about every benchmark. Ideally, for every new benchmark, we want to add only a new directory. No more, no less.

grief8 replied:
For benchmark preparation, we can remove it because now the `Makefile` is responsible for moving and compiling benchmarks.

I will also replace `benchmark/config.json` with `<benchmark>/threshold.json` for every benchmark.

tatetian commented on regression/benchmark/benchmarks.sh:
```diff
@@ -0,0 +1,125 @@
+#!/bin/bash
+
```
Rename this file to `run_all.sh`, which is more self-describing than `benchmarks.sh`.

tatetian commented:
The revised version looks much better. We are really close to merge this PR. I am very excited about it. I raised some extra issues as shown above.

grief8 commented on .github/workflows/benchmarks.yml:
```diff
@@ -0,0 +1,66 @@
+    - uses: actions/checkout@v2
+    - name: Set up the environment
+      run: |
+        if ! command -v jq &> /dev/null; then
```
It is already included in the Dockerfile in this PR. I will modify it.

grief8 commented on regression/benchmark/sysbench-cpu/result_template.json:
```diff
@@ -0,0 +1,14 @@
+[
+    {
+        "name": "Linux CPU Average Latency",
```
Given the following sysbench results, will `Linux CPU Average Test Latency` be better? We will indicate sysbench on the title of the chart.

```log
Latency (ms):
         min:                                    0.02
         avg:                                   17.03
         max:                                  258.26
         95th percentile:                       46.63
         sum:                             11999698.40
```

grief8 replied:
Change to `Average Execution Time per CPU on Linux`.

grief8 commented on regression/benchmark/getpid/run.sh:
```diff
@@ -0,0 +1,17 @@
+    fi
+fi
+
+/benchmark/getpid.o
```
Get it

grief8 commented on regression/benchmark/getpid/run.sh:
```diff
@@ -0,0 +1,17 @@
+# If not there, exit
+if [ ! -f /benchmark/getpid.o ]; then
+    if [ -f /regression/getpid/getpid ]; then
+        cp /regression/getpid/getpid /benchmark/getpid.o
```
I add `@gcc $(CUR_DIR)/apps/getpid/getpid.c -o $@/getpid` to generate getpid under `benchmark/bin` now.

grief8 commented:
TODO:

1. Add `self-host` runner.
2. Add a token `BENCHMARK_SECRET` to `asterinas`.

tatetian commented on regression/benchmark/run_all.sh:
```diff
@@ -0,0 +1,109 @@
+    local result_file="result_${benchmark}.json"
+
+    # Init script for initramfs
+    local init_script="${BENCHMARK_DIR}/init"
```
The first process is called the init process. But the program that is run as the init process should not be called the init program. When it is a script, it should not be called "init script". 

```
# Init script for initramfs
```

Reading this comment, I would think this is a script that initializes the content of the initramfs. To avoid such ambiguity, describe what the script does in its name.

grief8 replied:
I change it to `entrypoint` now. 
```sh
# Entrypoint script for initramfs
local initramfs_entrypoint_script="${BENCHMARK_DIR}/entrypoint.sh"
generate_entrypoint_script "${benchmark}" > "${initramfs_entrypoint_script}"
chmod +x "${initramfs_entrypoint_script}"
```

tatetian commented on regression/Makefile:
```diff
@@ -1,10 +1,12 @@
 BUILD_DIR := $(CUR_DIR)/build
-VDSO_DIR := $(BUILD_DIR)/linux_vdso
 INITRAMFS := $(BUILD_DIR)/initramfs
+INIT_FILE := $(CUR_DIR)/benchmark/init
```
This file name is inappropriate: the variable has very general name, `INIT_FILE`, but it only applies to benchmark.

grief8 replied:
Also rename them to `BENCHMARK_ENTRYPOINT` and `benchmark_entrypoint.sh`.

tatetian commented on regression/Makefile:
```diff
@@ -75,11 +85,15 @@ $(INITRAMFS)/usr/bin: | $(INITRAMFS)/bin
 	@cp /usr/local/benchmark/sysbench/bin/sysbench $@
 	@cp /usr/local/benchmark/iperf/bin/iperf3 $@
 	@cp /usr/local/benchmark/membench/membench $@
+	@gcc $(CUR_DIR)/apps/getpid/getpid.c -o $@/getpid
```
Add `-O2` flag.

And add a TODO.

```bash
# Replace the homebrewed getpid with a standard benchmark like UnixBench or LMbench
```

tatetian commented on regression/Makefile:
```diff
@@ -32,7 +35,13 @@ SYSCALL_TEST_DIR := $(INITRAMFS)/opt/syscall_test
+$(INITRAMFS):
+	@mkdir -p $@
+	@if [ -e $(INIT_FILE) ]; then \
+		cp $(INIT_FILE) $@; \
```
It is inappropriate to put the init script for benchmark under the root directory. From its name ("init"), one cannot tell that the file is for benchmark. Reorganize the code to make benchmark-related files to more proper locations.

grief8 replied:
Move it to `benchmark` now.

tatetian commented on tools/docker/Dockerfile.jinja:
```diff
@@ -16,10 +16,18 @@ RUN apt update && apt-get install -y --no-install-recommends \
+WORKDIR /root/dependency
+RUN wget https://raw.githubusercontent.com/asterinas/linux_vdso/2a6d2db/vdso64.so -O vdso64.so
+RUN wget https://raw.githubusercontent.com/asterinas/linux_kernel/9e66d28/vmlinuz-5.15.0-105-generic \
+        -O vmlinuz
```
Extract the modifications to the Dockerfile into another PR and bump the version to upload the new Docker image

grief8 replied:
Now #978 bumps the version.

tatetian replied:
#978 has been merged. Now remove the changes to the Dockerfile in this PR. And rebase to the latest main branch.

tatetian reviewed (approved):
LGTM. Thanks for the contribution!
